### ISSUE TITLE
gfal/davix parent creation fails with XRootD 5.0.1 in manager+DTN setup if client relies on session caching

### ISSUE THREAD

**Issue by olifre** [xrootd/xrootd#1276]
I'm unsure if this is some kind of follow-up for #1252, it is pretty confusing... 

Here's what I observe:
- `gfal-copy -p` copying to a non-existent directory runs fine against XRootD 4.12.x, i.e. the parent directories are created. 
- `gfal-copy -p` also runs fine against a standalone XRootD 5.0.1 server, parent directories are created. 
- `gfal-copy -p` **fails** to create parent directories and aborts when a redirecting manager (no matter if 4.12.x or 5.0.1) is combined with a 5.0.1 data node, unless `DAVIX_DISABLE_SESSION_CACHING=true` is set. 

The error on the client side is:
```
[gfalt_copy_file][perform_copy][gfal_http_copy] DESTINATION MAKE_PARENT [gfal_http_copy_make_parent]   [gfal_http_exists][gfal_http_access][davix2gliberr] Result Authentication error, reached maximum number of attempts after 1 attempts
```

That's pretty confusing, here some more details. This is the failing case:
```
$ gfal-copy -p -vvv -f -K adler32 --checksum-mode both davs://somewhere.example.com/something.root davs://managernode.example.com:1094/cephfs/grid/atlas/user/scratch/bla3/foo.root
...
DEBUG    Davix: > PROPFIND /cephfs/grid/atlas/user/scratch/bla3/foo.root HTTP/1.1
...
DEBUG    Davix: < HTTP/1.1 302 Redirect
...
DEBUG    Davix: < Location: https://dataserver.example.com:1094/cephfs/grid/atlas/user/scratch/bla3/foo.root
...
DEBUG    Davix: > PROPFIND /cephfs/grid/atlas/user/scratch/bla3/foo.root HTTP/1.1
...
> Host: dataserver.example.com:1094
...
DEBUG    Davix: < HTTP/1.1 404 Not Found
DEBUG    Davix: < Connection: Close
DEBUG    Davix: < Content-Length: 90
...
...
...
DEBUG    Davix: Found redirection <PROPFIND davs://managernode.example.com:1094/cephfs/grid/atlas/user/scratch/bla3/foo.root https://dataserver.example.com:1094/cephfs/grid/atlas/user/scratch/bla3/foo.root>
DEBUG    Davix:  creating session keys... httpsdataserver.example.com:1094
DEBUG    Davix: cached ne_session found ! taken from cache 
DEBUG    Davix: configure session...
DEBUG    Davix: define connection timeout to 30
DEBUG    Davix: define operation timeout to 8000
DEBUG    Davix: add CA PATH /etc/grid-security/certificates/
DEBUG    Davix: disable login/password authentication
DEBUG    Davix: enable client cert authentication by callback 
DEBUG    Davix:  -> negotiateRequest
DEBUG    Davix: NEON start internal request
DEBUG    Davix: Running pre_send hooks
DEBUG    Davix: > PROPFIND /cephfs/grid/atlas/user/scratch/bla3/foo.root HTTP/1.1
> User-Agent: gfal2-util/1.5.3 gfal2/2.18.1 neon/0.0.29
> TE: trailers
> Host: dataserver.example.com:1094
> Depth: 0
> 

DEBUG    Davix: Sending request-line and headers:
DEBUG    Davix: Doing SSL negotiation.
DEBUG    Davix: Request sent; retry is 0.
DEBUG    Davix: Aborted request (-4): Could not read status line
DEBUG    Davix: sess: Closing connection.
DEBUG    Davix: sess: Connection closed.
DEBUG    Davix: Delete Cached redirection for <PROPFIND davs://managernode.example.com:1094/cephfs/grid/atlas/user/scratch/bla3/foo.root https://dataserver.example.com:1094/cephfs/grid/atlas/user/scratch/bla3/foo.root>
DEBUG    Davix: Disable Session recycling
DEBUG    Davix:  ->  Error when using reycling of session/redirect : cancel and try again
INFO     Davix: Connection problem, retry
DEBUG    Davix: (EndRequest)(Libneon) Operation incomplete, kill the connection
DEBUG    Davix: sess: Not closing closed connection.
DEBUG    Davix: Connection problem: eradicate session
DEBUG    Davix: Disable Session recycling
DEBUG    Davix: Running post_send hooks
DEBUG    Davix: sess: Not closing closed connection.
DEBUG    Davix: Running destroy hooks.
DEBUG    Davix: Request ends.
```

After this, it tries to recover with metalink, and expectedly fails. It seems `davix` does not re-attempt a connection to the dataserver node without the session cache usage. 

Things work well when I use this straightaway:
```
export DAVIX_DISABLE_SESSION_CACHING=true
```

I tried both:
```
http.tlsreuse off
```
and
```
http.tlsreuse on
```
on the dataserver node, no change. Running `gfal-copy` direclty against the data server also works. 

Logs of the dataserver are pretty harmless:
```
200831 22:52:36 3487417 XrootdBridge: eyermuth.28:29@lxplus7118.cern.ch login as eyermuth
200831 22:52:36 3487417 acc_Audit: eyermuth.28:29@lxplus7118.cern.ch grant gsi eyermuth@[::ffff:188.184.103.253] stat /cephfs/grid/atlas/user/scratch/bla3/foo.root
200831 22:52:36 3487417 ofs_stat: eyermuth.28:29@lxplus7118.cern.ch Unable to locate /cephfs/grid/atlas/user/scratch/bla3/foo.root; no such file or directory
200831 22:52:36 3487417 XrootdXeq: eyermuth.28:29@lxplus7118.cern.ch disc 0:00:00 (send failure)
200831 22:52:37 3487243 XrootdBridge: eyermuth.29:29@lxplus7118.cern.ch login as eyermuth
200831 22:52:37 3487243 acc_Audit: eyermuth.29:29@lxplus7118.cern.ch grant gsi eyermuth@[::ffff:188.184.103.253] stat /cephfs/grid/atlas/user/scratch/bla3/foo.root
200831 22:52:37 3487243 ofs_stat: eyermuth.29:29@lxplus7118.cern.ch Unable to locate /cephfs/grid/atlas/user/scratch/bla3/foo.root; no such file or directory
200831 22:52:37 3487243 XrootdXeq: eyermuth.29:29@lxplus7118.cern.ch disc 0:00:00 (send failure)
```
Checking only the failed connection at highest debug level shows:
```
200831 22:57:16 3487721 XrdInet: Accepted connection from 32@lxplus7118.cern.ch
200831 22:57:16 3487700 XrdSched: running main accept inq=0
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: received dlen: 16
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: received dump: 22 03 01 19 17 01 00 19 13 03 03 97 115 75 73 00 
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: This does not look like http at pos 0
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: This may look like https
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: Protocol matched. https: 1
200831 22:57:16 3487721 XrdProtLoad: matched port 1094 protocol XrdHttp
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch XrdPoll: FD 32 attached to poller 1; num=1
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp:  Process. lp:0x7f9448006ba8 reqstate: 0
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp:  Setting host: [::ffff:188.184.103.253]
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp:  Entering SSL_accept...
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp:  SSL_accept returned :1
200831 22:57:16 3487721 cryptossl_X509::CertType: certificate has 3 extensions
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: Failed to extract DN information.
200831 22:57:16 3487721 sysXrdHttp:  Cleanup
200831 22:57:16 3487721 sysXrdHttp:  Reset
200831 22:57:16 3487721 sysXrdHttp:  XrdHttpReq request ended.
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch XrdPoll: Poller 1 removing FD 32
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch XrdPoll: FD 32 detached from poller 1; num=0
```

In short, I am unsure what exactly is happening — things only break when using both a redirector and a dataserver node, only if the client uses session caching, and no matter if XRootD uses session caching, but the same setup works with 4.12.x. 
Sadly, this means that it fails when FTS tries to transfer data.

**Comment by abh3** [comment]
So, let me get this correctly. When a redirector is in the loop, it fails when redirected to a 5.x data server but works when redirected to a <= 4.12.3 server. Is that correct?

In any case, it seems the issue is here:
200831 22:57:16 3487721 cryptossl_X509::CertType: certificate has 3 extensions
200831 22:57:16 3487721 anon.0:32@lxplus7118.cern.ch sysXrdHttp: Failed to extract DN information.

While the data server gets the cert it can't extract the DN. This particular error occurs in code recently added to handle certs in the "proper" way.  In this case, a DN is required for anticipated use with a gridmap file. It would seem you would get the same problem with a 4.12.4-RCx data server which uses the same code to do this. Perhaps you can try that to verify that it's a problem in this new code. 

Unfortunately, the newly added code does not propagate the debug settings to the underlying x509 library. You can force this by configuring gsi in xrootd with the proper debug flags set there.

**Comment by olifre** [comment]
> When a redirector is in the loop, it fails when redirected to a 5.x data server but works when redirected to a <= 4.12.3 server. Is that correct?

Exactly, that's the only way I can reproduce it. Of course, it does not mean that this is the cause — it could be that this triggers the correct "timing" / delays between reusing cached sessions to cause this. 



> It would seem you would get the same problem with a 4.12.4-RCx data server which uses the same code to do this. Perhaps you can try that to verify that it's a problem in this new code.

I will try. I can not promise a timeline yet, though: The debug information was collected in our production setup after an upgrade with fingers crossed, since I was unable to see this issue with our test setup, since our test setup currently is a single node only. So I will first have to expand our test setup to two nodes to freely test also RCs there, I did not anticipate there would be issues only visible with a redirector, but of course issues always hide in the single difference you have between test and prod ;-). 



> Unfortunately, the newly added code does not propagate the debug settings to the underlying x509 library. You can force this by configuring gsi in xrootd with the proper debug flags set there.

You mean the `-d` setting here?
https://xrootd.slac.stanford.edu/doc/dev50/sec_config.htm#_Toc42289901
I'll set these after reproducing this in our test environment (will take at least two days though to get a new Grid certificate for the node).

**Comment by abh3** [comment]
Thanks. Yes, the '-d:3' option, in this case. This will also dump the certs so something to only run for test mode. Instead of getting another node you could run two instances on the same node and use a different port number. That should work.

**Comment by olifre** [comment]
> Instead of getting another node you could run two instances on the same node and use a different port number. That should work.

Indeed, that also reproduces the issue :-). 
Here's what I get with 5.0.1 using `-d:3` for GSI:
```
200901 17:32:47 3034423 cryptossl_X509::CertType: certificate has 3 extensions
200901 17:32:47 3034423 cryptossl_X509::CertType: Found RFC 382{0,1}compliant proxyCertInfo extension
200901 17:32:47 3034423 crypto_X509Chain::EECname: EEC not found in chain
```
I'll install 4.12.4-RCx on the test node next and see if the issue also appears there.

**Comment by olifre** [comment]
I can't reproduce with 4.12.4-0.rc2 with exactly identical configuration... 

Let me know if you want additional logs (and at which debug level, and from client or server side).

**Comment by abh3** [comment]
So, 5.0.1 can't find the EEC but 4.12.4 can. So, the backport isn't 
identical. OK, how about giving me the log with gsi dump enabled for both 
runs (i.e. 4 vs 5) and maybe something will stick out.

**Comment by olifre** [comment]
Here's the collection of logs. I also added the client-side logs (`gfal-copy` with `-vvv`) in case correlating with the client-side requests helps to pinpoint it:, and have taken the logs once with `http.tlsreuse on` and once with `http.tlsreuse on` in case of XRootD 5.0.1. For gfal, settings were always the default (i.e. session reuse on). 

| Setup | GFAL log | XRootD log of DTN with `-d:3` for GSI |
|----------|---------------|--------------------------------------------------|
| XRootD 5.0.1, `http.tlsreuse on` | [gfal-xrd5-reuse.log](https://github.com/xrootd/xrootd/files/5158670/gfal-xrd5-reuse.log) | [xrd5-reuse.log](https://github.com/xrootd/xrootd/files/5158682/xrd5-reuse.log) |
| XRootD 5.0.1, `http.tlsreuse off` | [gfal-xrd5-noreuse.log](https://github.com/xrootd/xrootd/files/5158675/gfal-xrd5-noreuse.log) | [xrd5-noreuse.log](https://github.com/xrootd/xrootd/files/5158689/xrd5-noreuse.log) |
| XRootD 4.12.4-0.rc2 | [gfal-xrd4.log](https://github.com/xrootd/xrootd/files/5158678/gfal-xrd4.log) | [xrd4.log](https://github.com/xrootd/xrootd/files/5158691/xrd4.log) |

Let's hope something sticks out to your eyes :-).

**Comment by abh3** [comment]
So, here is what I found: 
1) While the tlsreuse directive was added no code ever tests the setting. That means the session cache is always enabled in R5, as it supposedly is in R4. So, no need to run two R5 tests until that gets fixed.
2) The failures in R5 are not consistent. The first "login" received a 3 extension cert (presumably the peer cert) followed by a 9 extension cert (presumably the cert chain). Some number of subsequent logins only receive the 3 extension cert and this is followed by an EEC failure. Then the whole thing recovers and you get the correct sequence. Then another failure. R4 does not exhibit this behavior.
3) There is not much difference between the R4 and R5 code paths. The one major difference is that R5 tests for refcount failure while R4 does not. That's a major oversight in R4 though one highly unlikely to ever cause a problem.


So, it would seem there is some strange interaction between davix and getting redirected to R5 or R4 server which would indicate there really is some fine difference we don't understand. So, let's nail down some more things:
a) What version is the redirector?
b) What OpenSSL version is being used by the R5 server vs the R4 server? The easiest way to find out is to do an lsof on each running program and looking to see what OpenSSL library is open. If they are different maybe that is the issue. Please note that if R4 is using OpenSSl version < 1.02, elliptic ciphers are not enabled.

So, relative to (b) you can test the hypothesis that it's the elliptic ciphers that are causing the problem by including the directive in the R5 server:
http.cipherfilter ALL:!LOW:!EXP:!MD5:!MD2

In the end I need to fix the tlsreuse option and instrument the whole thing to produce better messages. I can do that but it would mean you would need to run from source. Is that OK with you?

**Comment by bbockelm** [comment]
Hi,

This looks strongly like TLS session reuse.

When a SSL session is successfully reused, OpenSSL provides the client certificate but not the certificate chain.  If you are using a proxy, the client certificate is the proxy - and, as this is not an EEC, the DN extraction fails.

Oliver: you can probably confirm this by trying to trick GFAL to using your client certificate and not a proxy.

So, the question is: why is SSL reuse occurring?

Brian

**Comment by olifre** [comment]
Hi,

let's start from the back ;-). 

> Oliver: you can probably confirm this by trying to trick GFAL to using your client certificate and not a proxy.

indeed, that works!



> What version is the redirector?

In the logs I provided, both instances ran on the same machine, and both used the 5.0.1 binaries. That was on CentOS 8, i.e. openssl-1.1.1c. 
However, the earlier tests showing the same issue also covered the combinations:
- Redirector running 4.12.3 with OpenSSL 1.0.2k (CentOS 7), data server running 5.0.1 with OpenSSL 1.0.2k (CentOS 7). 
- Both the redirector and data server running 5.0.1 with OpenSSL 1.0.2k (CentOS 7). 

Both were affected, too. The error "looked the same", I could of course also try to catch logs, but that would take a bit (would then reproduce in our test setup). 



> So, relative to (b) you can test the hypothesis that it's the elliptic ciphers that are causing the problem by including the directive in the R5 server:  {{http.cipherfilter ALL:!LOW:!EXP:!MD5:!MD2}}

Since it also happened in the 5.0.1 <=> 5.0.1 combination on a single node with two instances, this is probably nought — I still tried it, but the error remains.

**Comment by bbockelm** [comment]
@olifre - I don't think there's much sense in running further experiments.  As @abh3 notes in his last comment, if you look at the `tlsreuse` patch, there is this line:

https://github.com/xrootd/xrootd/blob/master/src/XrdHttp/XrdHttpProtocol.cc#L1648

That is, regardless the value of `http.tlsreuse`, it is ignored as the session cache is hardcode to "on".

Knowing that session reuse is enabled when XRootD is unable to handle session reuse for grid proxies, the rest of the error messages make perfect sense.

One item though:

> That means the session cache is always enabled in R5, as it supposedly is in R4

I think the latter half of the statement is backwards.  The session cache is hardcoded off in R4 and there's no way to reenable it:

https://github.com/xrootd/xrootd/blob/stable-4.12.x/src/XrdHttp/XrdHttpProtocol.cc#L1686-L1691

**Comment by olifre** [comment]
> Knowing that session reuse is enabled when XRootD is unable to handle session reuse for grid proxies, the rest of the error messages make perfect sense.

Thanks for pointing this out (and the place in the code), then this completely matches all results from the experiments :-). 



> I think the latter half of the statement is backwards. The session cache is hardcoded off in R4 and there's no way to reenable it:

And that also matches observation, and explains why also 4.12.4-rc2 is unaffected.

**Comment by abh3** [comment]
Ah, correct! Indeed, there was coded added later on that shut off the session cache. Just found it. So, I'll keep this open until I push the fix to allow you to turn off the cache. Because we see this problem, I am going to make "off" the default. We don't need to rehash this next year after everyone forgot the problem. BTW can you do a source-based test after I push to master?

**Comment by olifre** [comment]
> BTW can you do a source-based test after I push to master?

Sure, I can :-).

**Comment by abh3** [comment]
OK, I pushed the fixes to master. Please tell me if they fix your problem and if so, I will close the ticket.

**Comment by olifre** [comment]
@abh3 Perfect, that fixes things!
For reference, I did the fill series of tests with master on CentOS 8 now:
- if I do not specify `http.tlsreuse` on any instance (redirector, DTN), it works. 
- If I specify `http.tlsreuse off` on both, it still works (expectedly). 
- if I specify `http.tlsreuse on` on the redirector only, I reproduce an authentication error when `gfal` tries to `stat`, asking the redirector. 
- if I specify `http.tlsreuse on` on the DTN only, I reproduce the authentication error with `gfal` on `DESTINATION MAKE_PARENT` which was the starting point of all this. 
- if I specify `http.tlsreuse on` on both, I reproduce the authentication error with `gfal` on `DESTINATION MAKE_PARENT` again. 

So in short: The new default fixes the issue, and using `http.tlsreuse on` anywhere breaks things. 
I think that is the last issue blocking our migration of the production instance to R5, at least I believe we can only find further issues (if there are any) in production, so I'll go for R5 again when 5.0.2 comes out :-).

**Comment by abh3** [comment]
Great! Thanks again for all the testing. I'll close this issue then.
